internal/waitext/extensions.go

```diff
@@ -0,0 +1,153 @@
+// This backoff will run "forever", use WithMaxRetries or a context to put a hard cap.
+// This works similar to ConstantBackOff in github.com/cenkalti/backoff
+func NewConstantBackoff(interval time.Duration) wait.Backoff {
+	return wait.Backoff{
+		Duration: interval,
+		Steps:    waitPkgForeverSteps,
+	}
+}
+
+// WithMaxRetries creates a new backoff that has all the same settings as the input except for backoff.Steps
+// This will cause it to retry up to value of times when passed to wait.ExponentialBackoff
+// Combine with RetryWithContext or Retry
+func WithMaxRetries(backoff wait.Backoff, times int) wait.Backoff {
+	return wait.Backoff{
+		Duration: backoff.Duration,
+		Factor:   backoff.Factor,
+		Jitter:   backoff.Jitter,
+		Steps:    times + 1, // Initial execution should not count as retry so we add it as a step
+		Cap:      backoff.Cap,
+	}
+}
+
+// WithJitter will do randomization on every step on the backoff, causing the wait value to be in [duration, duration+jitter*duration]
+// Use when many clients could use be calling the same operation concurrently as this spreads out the calls a bit instead of converging on the same value
+func WithJitter(backoff wait.Backoff, randomizationFactor float64) wait.Backoff {
+	return wait.Backoff{
+		Duration: backoff.Duration,
+		Factor:   backoff.Factor,
+		Jitter:   randomizationFactor,
+		Steps:    backoff.Steps,
+		Cap:      backoff.Cap,
+	}
+}
+
+// Retry acts as RetryWithContext but with context.Background()
+func Retry(backoff wait.Backoff, operation func() error, errNotify func(error)) error {
+	return retryCore(context.Background(), backoff, func(_ context.Context) error {
+		return operation()
+	}, errNotify)
+}
+
+// RetryWithContext executes an operation with retries following these semantics:
+//
+//   - The operation is executed at least once (even if context is cancelled)
+//
+//   - If operation returns an error that is _not_ NonTransientError, the operation might be retried (see below for more info when)
+//
+//   - If operation returns an error that is NonTransientError, the operation is not retried and underlying error is unwrapped
+//
+// The operation will not be retried anymore if
+//
+//   - backoff.Steps reaches 0
+//
+//   - the context is cancelled
+//
+// The end result is the final error observed when calling operation() or nil if successful or context.Err() if the context was cancelled.
+// If retryNotify is passed, it is called when making retries.
+// Caveat: this function is similar to wait.ExponentialBackoff but has some important behavior differences like at-least-one execution and retryable errors
+func RetryWithContext(ctx context.Context, backoff wait.Backoff, operation func(context.Context) error, retryNotify func(error)) error {
+	return retryCore(ctx, backoff, operation, retryNotify)
+}
+
+func retryCore(ctx context.Context, backoff wait.Backoff, operation func(context.Context) error, retryNotify func(error)) error {
```

Comment:
Hey I can't stop but feel this is too complex. I think we can just have one retry functions and that would satisfy all use cases. Could you consider this?